rotkehlchen/balances/historical.py

```diff
@@ -1,29 +1,131 @@
 from rotkehlchen.assets.asset import Asset
-from rotkehlchen.constants import ZERO
+from rotkehlchen.constants import DAY_IN_SECONDS, ONE, ZERO
 from rotkehlchen.constants.prices import ZERO_PRICE
-from rotkehlchen.db.filtering import HistoryEventFilterQuery
-from rotkehlchen.db.history_events import DBHistoryEvents
+from rotkehlchen.db.filtering import (
+    DBFilter,
+    DBMultiStringFilter,
+    DBNestedFilter,
+    DBNotEqualFilter,
+    DBTimestampFilter,
+)
 from rotkehlchen.errors.misc import NotFoundError, RemoteError
 from rotkehlchen.errors.price import NoPriceForGivenTimestamp
 from rotkehlchen.fval import FVal
-from rotkehlchen.history.events.structures.types import EventDirection
+from rotkehlchen.globaldb.handler import GlobalDBHandler
+from rotkehlchen.history.events.structures.types import (
+    EventDirection,
+    HistoryEventSubType,
+    HistoryEventType,
+)
 from rotkehlchen.history.price import PriceHistorian
-from rotkehlchen.types import Timestamp
-from rotkehlchen.utils.misc import ts_ms_to_sec
+from rotkehlchen.types import Timestamp, TradeType
+from rotkehlchen.utils.misc import timestamp_to_daystart_timestamp
 
 if TYPE_CHECKING:
     from rotkehlchen.db.dbhandler import DBHandler
-    from rotkehlchen.history.events.structures.base import HistoryBaseEntry
 
 
 class HistoricalBalance(TypedDict):
     amount: FVal
     price: FVal
 
 
+@dataclass(init=True, repr=True, eq=True, order=False, unsafe_hash=False, frozen=True)
+class MinimalTrade:
+    timestamp: Timestamp
+    identifier: str
+    amount: FVal
+    trade_type: TradeType
+    base_asset: Asset
+    quote_asset: Asset
+    rate: FVal
+
+    def update_balances(self, current_balances: dict[Asset, FVal]) -> str | None:
+        """Updates current balances for a trade event, checking for negative balances.
+
+        Returns the event_identifier if a negative balance would occur, else None.
+        For trades:
+        - BUY means we receive base_asset and spend quote_asset
+        - SELL means we spend base_asset and receive quote_asset
+        """
+        if self.trade_type in {TradeType.BUY, TradeType.SETTLEMENT_BUY}:
+            quote_amount = self.amount * self.rate
+            if current_balances[self.quote_asset] - quote_amount < ZERO:
+                return self.identifier
+
+            current_balances[self.base_asset] += self.amount
+            current_balances[self.quote_asset] -= quote_amount
+            if current_balances[self.quote_asset] == ZERO:
+                del current_balances[self.quote_asset]
+        else:
+            if current_balances[self.base_asset] - self.amount < ZERO:
+                return self.identifier
+
+            current_balances[self.base_asset] -= self.amount
+            current_balances[self.quote_asset] += self.amount * self.rate
+            if current_balances[self.base_asset] == ZERO:
+                del current_balances[self.base_asset]
```

Comment:
this two look really similar. Isn't it possible to make it a single function method that you call passing the arguments in the right order?